x-pack/filebeat/input/awss3/s3_input.go

```diff
@@ -115,8 +115,19 @@ func (in *s3PollerInput) runPoll(ctx context.Context) {
 	}
 
 	// Start reading data and wait for its processing to be done
-	in.readerLoop(ctx, workChan)
+	ids, ok := in.readerLoop(ctx, workChan)
 	workerWg.Wait()
+
+	if !ok {
+		in.log.Warn("skipping state registry cleanup as object reading ended with a non-ok return")
+		return
+	}
+
+	// Perform state cleanup operation
+	err := in.states.CleanUp(ids)
```

Comment:
This looks like it doesn't handle the case where the bucket list prefix is nonempty -- if a configuration tries reading from only a specific prefix, then any registry data without that prefix will be purged, even if it's still in the bucket.

Reply:
@faec sorry, but I don't see any issue here. If `bucket_list_prefix` (which maps to `prefix`) is defined, s3 list object returns objects matching the prefix. Once processed, we keep track of them in the registry. Eventually, if user decides to remove objects (for example, using a lifecycle policy), then the list object will not return the removed entry, where we trigger the removal for the entry we no longer see. 

Is there any other pitfall you see here ?

Reply:
If there are two inputs targeting two different prefixes in the same bucket, they will erase each other's data.

Reply:
@faec I opened https://github.com/elastic/beats/pull/41869 to first improve how s3 states store work. With the improvement, store only load and store entries matching the given prefix. Prefix here refers to `bucket_list_prefix` and this could be empty `""` which is the non-prefix scenario.

Reply:
@faec Thanks for the review on https://github.com/elastic/beats/pull/41869.  The cleanup should now be safe as input specific states are prefix aware.